sphinx/writers/latex.py

```diff
@@ -1991,14 +1992,15 @@ def visit_container(self, node):
             for id in self.pop_hyperlink_ids('code-block'):
                 ids += self.hypertarget(id, anchor=False)
             if node['ids']:
-                ids += self.hypertarget(node['ids'][0])
+                # suppress with anchor=False \phantomsection generation
+                ids += self.hypertarget(node['ids'][0], anchor=False)
             self.body.append('\n')
-            self.context.append(ids + '\n')
+            # context.pop will be done in visit_caption
+            self.context.append(ids)
```

Comment:
@tk0miya New version does not use context, but interacts via a LaTeX macro which is checked for by Verbatim. In this way I can make functional links of the `` :ref:`Go there <label>` `` type, with no caption. I discovered that "visit_container" is not gone through if no caption, hence moved to "visit_literal_block" the handling in case of no caption.

Here is my test file:

```rst
TEST
====

FIRST
-----

The :numref:`toolbox.code_sdo1` code demonstrates how to solve this problem.
See also :ref:`Here <foo1>`, :numref:`This is foo`,
:numref:`foo2`, and also :ref:`There <foo3>` and :ref:`foo4`.

.. literalinclude:: sdo1.m
   :name: toolbox.code_sdo1
   :caption: Code implementing problem

 
SECOND
------

This is standard code-block (no name no caption)::

>>> x=y

This is a literal include with name ``foo1`` but without caption.

.. literalinclude:: sdo1.m
   :name: foo1

..

   This should be indented. A literal include with caption but no name.

   .. literalinclude:: sdo1.m
      :caption: This is foo

   ..

      This one has both caption and name.

      .. literalinclude:: sdo1.m
         :name: foo2
         :caption: This is foo2

THIRD
-----

Here a code-block directive with name and without caption.

.. code-block:: python
   :name: foo3

   >>> Hello=['name but no caption']

Here a code-block directive with caption and without name.

.. code-block:: python
   :caption: foo4

   >>> Hello=['caption but no name']

Here is a code-block directive with neither caption nor name.

.. code-block:: python

   >>> Hello=['no caption no name']
```